gaia-header.js

```diff
@@ -34,105 +51,270 @@ module.exports = Component.register('gaia-header', {
    */
   attached: function() {
-    this.rerunFontFit();
+    debug('attached');
+    this.nextTick(() => this.runFontFit());
+    this.observerStart();
   },
 
   /**
-   * Called when one of the attributes
-   * on the element changes.
+   * Called when the element is
+   * detached from the DOM.
    *
    * @private
    */
-  attributeChanged: function(attr) {
-    if (attr === 'action') {
-      this.configureActionButton();
-      this.rerunFontFit();
-    }
+  detached: function() {
+    debug('detached');
+    this.observerStop();
   },
 
   /**
-   * Runs the logic to size and position
-   * header text inside the available space.
+   * Run the font-fit logic and
+   * center the title.
    *
-   * @private
+   * Function calls ordering and use of
+   * `requestAnimationFrame` critical
+   * to minimising unwanted reflows.
+   *
+   * @public
    */
   runFontFit: function() {
-    for (var i = 0; i < this.els.headings.length; i++) {
-      fontFit.reformatHeading(this.els.headings[i]);
-      fontFit.observeHeadingChanges(this.els.headings[i]);
-    }
+    debug('run font-fit');
+    if (this.noFontFit) { return; }
+    var space = this.getTitleSpace();
+
+    // Stop the mutation observer
+    // before making any DOM changes.
+    this.observerStop();
+
+    // Fit the text and then show each title.
+    // Once all titles have been fit, we can
+    // start observing DOM mutations again.
+    return Promise.all(map.call(this.els.titles, (el) => {
+      return this.fitTitleText(el, space);
+    })).then(() => { this.observerStart(); });
   },
 
   /**
-   * Rerun font-fit logic.
+   * Fit text and center a title between
+   * the buttons before and after.
    *
-   * TODO: We really need an official API for this.
+   * @param  {HTMLH1Element} title
+   * @param  {Number} space
+   * @return {Promise}
+   * @private
+   */
+  fitTitleText: function(title, space) {
+    return new Promise((resolve) => {
+      debug('fit title text');
+      var text = this.getTitleText(title);
+
+      // Bail when there's no text
+      if (!text.value) { return resolve(title); }
+
+      // If neither the text or the titleSpace
+      // changed, there's no reason to continue.
+      if (!text.changed && !space.changed) { return resolve(title); }
+
+      var marginStart = this.getTitleMarginStart();
+      var textSpace = space.value - Math.abs(marginStart);
+      var fontFitResult = this.fontFit(text.value, textSpace);
+      var overflowing = fontFitResult.textWidth > textSpace;
+
+      // If the text is overflowing in the
+      // centered title, we remove centering
+      // to free up space, rerun fontFit to
+      // get a fontSize white fits this space.
+      if (overflowing) {
+        fontFitResult = this.fontFit(text.value, space.value);
+        marginStart = 0;
+      }
+
+      // Defer dom writes. Observer
+      // is silenced to prevent mutation
+      // callbacks running font-fit again.
+      this.nextTick(() => {
```

Comment:
why are you defering dom writes? is it to prevent reflows if we need to get values from the DOM for other titles? Can you explain this in the comment?

the comment about the observer should be near the code that actually silences the observer (in runFontFit)